types/index.d.ts

```diff
@@ -1,18 +1,21 @@
 import {
   ContextConfigDefault,
-  RouteGenericInterface,
   FastifyInstance,
   FastifyPluginCallback,
   FastifyReply,
   FastifyRequest,
   FastifySchema,
+  RouteGenericInterface,
   preHandlerHookHandler
 } from 'fastify';
 
 declare module 'fastify' {
   interface FastifyInstance<RawServer, RawRequest, RawReply, Logger, TypeProvider> {
-    auth(
-      functions: fastifyAuth.FastifyAuthFunction[] | (fastifyAuth.FastifyAuthFunction | fastifyAuth.FastifyAuthFunction[])[],
+    auth<
+      Request extends FastifyRequest = FastifyRequest,
+      Reply extends FastifyReply = FastifyReply
```

Comment:
Hmm, looks not correct. Why is Request and Reply not depending on the Generic Types from FastifyInstance?

```suggestion
      Request extends FastifyRequest = Request extends FastifyRequest = FastifyRequest<RawServerDefault, RawRequest, FastifySchema, TypeProvider, ContextConfigDefault, Logger>,
      Reply extends FastifyReply = FastifyReply<RawServer, RawRequest, RawReply, RouteGenericInterface, ContextConfigDefault, FastifySchema, TypeProvider>
```

Untested but I would expect this.

Reply:
The `FastifyRequest` and `FastifyReply` interfaces have changed from your example.
The default should be equal to the old value to prevent breaking the previous code.
Furthermore,  now the `FastifyRequest` accepts other generic types, for instance, RouteGeneric, and in this way, the users can pass the request as they prefer.
If you take a look at the tests, I tried to break nothing and include your request to leave the power to the user to choose the request type.